test/unit/BPool.t.sol

```diff
@@ -0,0 +1,557 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
+
+import {BConst} from 'contracts/BConst.sol';
+import {BPool} from 'contracts/BPool.sol';
+import {IERC20} from 'contracts/BToken.sol';
+import {Test, console} from 'forge-std/Test.sol';
+import {LibString} from 'solmate/utils/LibString.sol';
+
+abstract contract Base is Test, BConst {
+  using LibString for *;
+
+  BPool public bPool;
+  address[8] public tokens;
+
+  function setUp() public {
+    bPool = new BPool();
+
+    // Create fake tokens
+    for (uint256 i = 0; i < tokens.length; i++) {
+      tokens[i] = makeAddr(i.toString());
+    }
+  }
+
+  function _preparePool(
+    BPool _pool,
+    address[8] memory _tokens,
+    uint256[8] memory _balance,
+    uint256[8] memory _denorm
+  ) internal {
+    // Create mocks
+    for (uint256 i = 0; i < _tokens.length; i++) {
+      vm.mockCall(_tokens[i], abi.encodeWithSelector(IERC20(_tokens[i]).transfer.selector), abi.encode(true));
+      vm.mockCall(_tokens[i], abi.encodeWithSelector(IERC20(_tokens[i]).transferFrom.selector), abi.encode(true));
+    }
+
+    // Set tokens
+    for (uint256 i = 0; i < _tokens.length; i++) {
```

Comment:
can we just modify the storage?

lets try to make as many internals here as we can, remembering that each happy path for each method may need different methods being called

we try to avoid, unless not possible, calling any method on the pool, as is the contract we're testing